File: src/Like.java
public class Like {

    /**
     * Visual Basic's Like operator in Java.<br />
     * <br />
     * <code>?</code> matches exactly one character<br />
     * <code>*</code> matches zero or more characters<br />
     * <code>#</code> matches exactly one digit<br />
     * 
     * @author Urs Schmidt
     */
    public static boolean like(String str, String pattern) {
        final int sl = str.length();
        final int pl = pattern.length();
        for (int i = 0; i < pl; i++) {
            final char pc = pattern.charAt(i);
            if (pc != '*' && i == sl)
                return false;
            switch (pc) {
            case '?':
                /* always matches */
                break;
            case '*':
                if (i + 1 == pl)
                    return true;
                final String ps = pattern.substring(i + 1);
                for (int j = i; j < sl; j++) {
                    if (like(str.substring(j), ps))
                        return true;
                }
                return false;
            case '#':
                if (!Character.isDigit(str.charAt(i)))
                    return false;
                break;
            default:
                if (pc != str.charAt(i))
                    return false;
            }
        }
        return sl == pl;
    }

    public static void main(String[] args) {
        if (args.length < 2) {
            System.err.println("Usage: java Like str pattern");
            return;
        }

        final String str = args[0];
        final String pattern = args[1];
        final boolean retval = like(str, pattern);
        System.out.println(retval ? "true" : "false");
    }

}

File: src/like.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

/**
 *  Visual Basic's Like operator in C.
 *
 *  ? matches exactly one character
 *  * matches zero or more characters
 *  # matches exactly one digit
 *
 *  @author Urs Schmidt
 */
bool like(const char *str, const char *pattern) {
    for (size_t i = 0;; i++) {
        const char sc = str[i], pc = pattern[i];
        switch (pc) {
        case 0:
            return sc == 0;
        case '?':
            if (!sc)
                return false;
            break;
        case '*':
            if (!pattern[i + 1])
                return true;
            for (size_t j = i; str[j]; j++) {
                if (like(&str[j], &pattern[i + 1]))
                    return true;
            }
            return false;
        case '#':
            if (!sc || !('0' <= sc && sc <= '9'))
                return false;
            break;
        default:
            if (sc != pc)
                return false;
        }
    }
    /* unreachable */
}

int main(int argc, char **argv) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s str pattern\n", argv[0]);
        return EXIT_FAILURE;
    }

    const char *str = argv[1];
    const char *pattern = argv[2];
    const bool retval = like(str, pattern);
    printf("%s\n", retval ? "true" : "false");

    return EXIT_SUCCESS;
}
